### PR TITLE
Feature/add pandas1 dtypes

### PR DESCRIPTION
**Pull request by The-Fonz**
This PR adds the capability to both write and read Pandas NA-dtypes. Writing Panda's Int types was already implemented, now we can also convert to those on reading, and writing/reading the new Pandas 1.0 datatypes is now also possible.

Note: Tests are working but it's still a bit messy. Should be cleaned up, but wanted to get this PR out to show the concept and get some feedback.

## The basic idea
Please see the updated README for basic usage.

Writing the Pandas Boolean, String and Int datatypes is not that hard: just add them to the dictionary that maps python/numpy/pandas types to Avro types. We do need a bit of extra conversion for the Boolean, Int types as fastavro cannot write the `pd.NA` value, so we convert right before writing. But that's basically all there is to it.

Reading is another story, as we can suddenly read either as "old" numpy datatypes or as "new". To toggle between these two option, the flag `na_dtypes` is added to `read_avro()`. We can do this by going through the Avro schema and seeing where we can apply this inverse mapping. We need to do this after loading the dataframe with Pandas, as it by default loads everything as "old" datatypes. So we are just converting dtypes after the fact based on the Avro schema. This allows us to avoid the nondeterministic nature of [df.convert_dtypes()](https://pandas.pydata.org/docs/whatsnew/v1.0.0.html#convert-dtypes-method-to-ease-use-of-supported-extension-dtypes). This is not something that can easily be implemented after loading outside of Pandavro, as we need the Avro schema to do this deterministic mapping.

### CODE REVIEW

File: pandavro/__init__.py [hunk 1]
```diff
@@ -94,9 +127,34 @@ def __schema_infer(df, times_as_micros):
...
+        if type(typelist) == str:
+            return typelist
+        # If a logical type dict, it has a type attribute. Return None as we don't touch logical types
+        elif type(typelist) == dict:
+            return None
+            return _filter(typelist["type"])
+        else:
+            l = [t for t in typelist if t != "null"]
+            if len(l) > 1:
+                raise ValueError(f"More items in list than 1. {l}")
```
Comment (ynqa) [inline pandavro/__init__.py:145]: This line was failed in Python 2.7, 3.5. f-strings was introduced on 3.6.

```
==================================== ERRORS ====================================
___________________ ERROR collecting tests/pandavro_test.py ____________________
../../../virtualenv/python3.5.6/lib/python3.5/site-packages/_pytest/python.py:445: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
../../../virtualenv/python3.5.6/lib/python3.5/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:968: in _find_and_load
    ???
<frozen importlib._bootstrap>:957: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:664: in _load_unlocked
    ???
<frozen importlib._bootstrap>:634: in _load_backward_compatible
    ???
../../../virtualenv/python3.5.6/lib/python3.5/site-packages/_pytest/assertion/rewrite.py:308: in load_module
    six.exec_(co, mod.__dict__)
tests/pandavro_test.py:4: in <module>
    import pandavro as pdx
E     File "/home/travis/build/ynqa/pandavro/pandavro/__init__.py", line 145
E       raise ValueError(f"More items in list than 1. {l}")
E                                                        ^
E   SyntaxError: invalid syntax
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.51 seconds ============================
```

Comment (The-Fonz) [reply pandavro/__init__.py]: Removed this f-string

**Comment by The-Fonz** [comment]
@ynqa I've pushed some fixes and cleanups. Also extended the tests for ints, with Pandas NA-dtypes we can do a full roundtrip but with the numpy (u)int dtypes fastavro reads them back as int64 (existing behavior). This was not fully tested before.

Tests for Python I still have to do some work on Pandas <1.0.0 support as I'm using `pd.NA` in the checks.
- [ ] Do not execute check with `pd.NA` if not available
- [ ] Do not run Pandas 1.0.0 tests if on previous version

Would it be an idea to add different Pandas version tests to Travis config?

File: README.md [hunk 2]
```diff
@@ -72,12 +92,17 @@ OUTPUT_PATH='{}/example.avro'.format(os.path.dirname(__file__))
...
+    df = pd.DataFrame({
+        "Boolean": [True, False, True, False],
+        "pdBoolean": pd.Series([True, False, True, False, True, None, True, False], dtype=pd.BooleanDtype()),
+        "Float64": np.random.randn(4),
+        "Int64": np.random.randint(0, 10, 4),
+        "pdInt64":  pd.Series(list(np.random.randint(0, 10, 7)) + [None], dtype=pd.Int64Dtype()),
+        "String": ['foo', 'bar', 'foo', 'bar'],
+        "pdString": pd.Series(['foo', 'bar', 'foo', 'bar', 'foo', 'bar', 'foo', 'bar'], dtype=pd.StringDtype()),
+        "DateTime64": [pd.Timestamp('20190101'), pd.Timestamp('20190102'),
+                       pd.Timestamp('20190103'), pd.Timestamp('20190104')]
```
Comment (YasenPetrov) [inline README.md:104]: Some of these arrays have 4 values and some 8. And the `pdString` column does not have any `None` values. This fixes these points:
```suggestion
        "Boolean": [True, False, True, False],
        "pdBoolean": pd.Series([True, None, True, False], dtype=pd.BooleanDtype()),
        "Float64": np.random.randn(4),
        "Int64": np.random.randint(0, 10, 4),
        "pdInt64":  pd.Series(list(np.random.randint(0, 10, 3)) + [None], dtype=pd.Int64Dtype()),
        "String": ['foo', 'bar', 'foo', 'bar'],
        "pdString": pd.Series(['foo', 'bar', 'foo', None], dtype=pd.StringDtype()),
        "DateTime64": [pd.Timestamp('20190101'), pd.Timestamp('20190102'),
                       pd.Timestamp('20190103'), pd.Timestamp('20190104')]
```

File: pandavro/__init__.py [hunk 3]
```diff
@@ -29,28 +34,47 @@
...
+
+    logger.debug("Imported pandas >=1.0.0 datatypes")
+except AttributeError:
+    logger.debug("Did not import pandas >=1.0.0 datatypes")
 
 
 def __type_infer(t):
     # Binary data has to be handled separately from the other dtypes because it
     # requires a parameter, the buffer size.
+    print(t)
```
Comment (YasenPetrov) [inline pandavro/__init__.py:77]: Stray print statement, I think.

```suggestion
```

**Comment by The-Fonz** [comment]
Hi @ynqa, is there any chance to get this merged? I see that some test are failing, would be willing to spend some time fixing that if merge is likely.

**Comment by marctorsoc** [comment]
Hi @The-Fonz , what happened with this? I'd be interested on getting this merged, if you can 🙏

**Comment by jeff-99** [comment]
@ynqa @YasenPetrov, I would love if this could get merged.

**Comment by marcglobality** [comment]
@The-Fonz it seems like cutting some of the contributions we could get some of this merged?

**Comment by The-Fonz** [comment]
I didn't look at this for some time as it was just humming along in production on our side. 
But thanks @ynqa for fixing and merging! I see it was released as v1.7.0, that's great.

**Pull request merged**